mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -138,48 +135,116 @@ static Value applyTransforms(PatternRewriter &b, ThreadwiseWriteAllOp storeOp,
   return alloc;
 }
 
-static Operation *traceToNonViewOp(Operation *op) {
+static void traceToNonViewUsers(Operation *op,
+                                SmallVectorImpl<Operation *> &nonViewOps) {
   if (auto transform = dyn_cast<TransformOp>(op)) {
     Value result = transform.getResult();
-    // TODO(sjw): fix when divergence is encountered
-    assert(result.hasOneUse());
     for (auto &use : result.getUses()) {
-      return traceToNonViewOp(use.getOwner());
+      traceToNonViewUsers(use.getOwner(), nonViewOps);
     }
+  } else {
+    nonViewOps.push_back(op);
+  }
+}
+
+static Operation *traceToNonViewDef(Operation *op,
+                                    SmallVectorImpl<TransformMapAttr> &views,
+                                    Operation *terminator) {
+  if (op == terminator) {
+    return op;
+  }
+  if (auto transform = dyn_cast<TransformOp>(op)) {
+    Operation *nonViewDef = traceToNonViewDef(
+        transform.getViewSource().getDefiningOp(), views, terminator);
+    views.push_back(transform.getTransformAttr());
+    return nonViewDef;
   }
   return op;
 }
 
-static ThreadwiseWriteAllOp traceToThreadwiseWrite(Value inp) {
+static Operation *traceToNonViewDef(Operation *op) {
+  if (auto transform = dyn_cast<TransformOp>(op)) {
+    return traceToNonViewDef(transform.getViewSource().getDefiningOp());
+  } else {
+    return op;
+  }
+}
+
+// This function checks the results of the current operation
+// is copied to a block argument output, hence needs to be
+// preserved.
+static bool isLeadingToBlockArg(Operation *op) {
+  Operation *nonViewDef = traceToNonViewDef(op);
+  for (Value result : nonViewDef->getResults()) {
```

Comment:
Sure -- but dont we have copies always in fusion cases ? now that we dont have copy-opt

Reply:
Yeah but suppose, for example, that I've decided to omit that trailing copy, either because someone actually wrote a general copy opt in the future or because, for example, I'm manually crafting the `la.generic`s myself

Reply:
done!